chainer/functions/activation/tanh.py

```diff
@@ -24,37 +24,68 @@ def forward_cpu(self, x):
         y = utils.force_array(numpy.tanh(x[0]))
         self.retain_inputs(())
         self.retain_outputs((0,))
+        self._use_cudnn = False
         return y,
 
     def forward_gpu(self, x):
         if chainer.should_use_cudnn('==always') and x[0].flags.c_contiguous:
             y = cudnn.activation_forward(x[0], _mode)
+            self.retain_inputs((0,))
+            self._use_cudnn = True
         else:
             y = cuda.cupy.empty_like(x[0])
             cuda.cupy.tanh(x[0], out=y)
             self.retain_inputs(())
+            self._use_cudnn = False
 
         self.retain_outputs((0,))
         return y,
 
-    def backward_cpu(self, x, gy):
-        y = self.output_data[0]
+    def backward(self, indexes, grad_outputs):
+        if self._use_cudnn:
+            x = self.get_retained_inputs()[0].data
+        else:
+            x = None
+        y = self.get_retained_outputs()[0]
+        gy = grad_outputs[0]
+        return TanhGrad(x).apply((y, gy))
+
+
+class TanhGrad(function_node.FunctionNode):
+
+    def __init__(self, x):
+        super(TanhGrad, self).__init__()
+        # The original input `x` is only required for cuDNN.
+        # If it is None, this class does not use cuDNN.
+        self.x = x
+
+    def forward_cpu(self, inputs):
+        self.retain_inputs((0, 1))
+        y, gy = inputs
         one = y.dtype.type(1)
-        return utils.force_array(gy[0] * (one - y * y)),
+        return utils.force_array(gy * (one - y * y)),
 
-    def backward_gpu(self, x, gy):
-        y = self.output_data[0]
+    def forward_gpu(self, inputs):
+        self.retain_inputs((0, 1))
+        y, gy = inputs
         if (chainer.should_use_cudnn('==always') and
-                x[0] is not None and x[0].flags.c_contiguous and
-                gy[0].flags.c_contiguous):
-            gx = cudnn.activation_backward(x[0], y, gy[0], _mode)
+                self.x is not None and self.x.flags.c_contiguous and
```

Comment:
The check on c-contiguity of `x` looks redundant.

Reply:
Why is it redundant?